lib/QubitFlatfileImport.class.php

```diff
@@ -1416,28 +1416,45 @@ public static function fetchKeymapEntryBySourceAndTargetName($sourceId, $sourceN
 
     /**
      * Map a value to its corresponding term name then return the term ID
-     * corresponding to the term name.
+     * corresponding to the term name. This function will create a new term if
+     * not found in taxonomy.
      *
-     * @param string $description        description of subject (for error output)
-     * @param string $value              value that needs to be mapped to a term ID
-     * @param array  $valueToTermNameMap array mapping possible values to term names
-     * @param array  $terms              array mapping term IDs to term names
+     * @param string $taxonomyDescription description of subject (for error output)
+     * @param string $value               value that needs to be mapped to a term ID
+     * @param mixed  $culture             row culture
+     * @param array  $terms               array mapping term IDs to term names
+     * @param mixed  $taxonomyId          taxonomy id of term to be created
      *
      * @return int term ID
      */
-    public function translateNameToTermId($description, $value, $valueToTermNameMap, $terms)
+    public function createOrFetchTermIdFromName($taxonomyDescription, $value, $culture, $terms, $taxonomyId)
     {
-        if (isset($valueToTermNameMap[$value]) || 0 == count($valueToTermNameMap)) {
-            $termName = (count($valueToTermNameMap)) ? $valueToTermNameMap[$value] : $value;
+        if (empty($value)) {
+            return;
+        }
 
-            if (in_array($termName, $terms)) {
-                return array_search($termName, $terms);
-            }
+        $key = csvImportBaseTask::arraySearchCaseInsensitive(
+            $value,
+            $terms[$culture]
+        );
 
-            throw new sfException('Could not find "'.$termName.'" in '.$description.' terms array.');
-        } else {
-            throw new sfException('Could not find a way to handle '.$description.' value "'.$value.'".');
+        if (false === $key) {
+            echo "\nTerm {$value} not found in {$taxonomyDescription} taxonomy, creating it...\n";
+            QubitFlatfileImport::createTerm(
```

Comment:
I think this approach is too risky - if the `$terms` value doesn't match the expected data or array structure (e.g. it's an empty array or a null) it will result in either (a) an error, if `$terms[$culture]` doesn't exist, or (b) `$value` will be added to the database, without any check that it's not a duplicate.

Reply:
I think `QubitFlatfileImport::createTerm()` should be moved to `lib/model/QubitTerm.php` too, as it's directly related to manipulating the model IMHO.

Reply:
The $terms array lookup by taxonomy was implemented in the past to avoid going to the database for every term since looking up terms by name is expensive. I think we can rely on the terms array to be updated, but in looking at this again, I realize I've broken the update of the terms array (csvImportBaseTask::refreshTaxonomyTerms) when a new term is created - the update is not carried out back to the status array.  I may not be able to move this logic into the new QubitFlatfileImport::createOrFetchTermIdFromName() as I had planned.

Reply:
@sbreker I have a few comments on this.

1. This is a public method, even if the $terms argument is 100% correct all the time in this class, it may not be used correctly when externally called.
2. This is a really complicated and big class, and I think the probability is high of somebody passing the wrong `$terms` data into the method. All it would take is a copy & paste of an existing call and forgetting to change the `$self->status['actorTypes']` key.
3. The final database safety check for the term can happen only if the cached `$terms` lookup misses.  Most imports tend to use existing terms a lot, so the ratio of cache hits to fails should be pretty high.